packages/middleware/rate-limiter/src/sliding-window.ts

```diff
@@ -0,0 +1,125 @@
+ * Example of maximum 10 request every 2 minutes:
+ * ```typescript
+ * new SlidingWindow({ rateLimit: 10, samplingPeriod: "2 minutes" })
+ * ```
+ */
+export class SlidingWindowProvider {
+  private readonly slidingWindows: Map<string, SlidingWindow> = new Map()
+  private readonly rate: Rate
+  private readonly slotProvider: SlotProvider
+
+  constructor({ rate, slotProvider }: { rate: Rate | RateLiteral; slotProvider: SlotProvider }) {
+    this.rate = typeof rate === 'string' ? parseRate(rate) : rate
+    this.slotProvider = slotProvider
+  }
+
+  get({ key }: { key: string }): SlidingWindow {
+    const slidingWindow = this.slidingWindows.get(key)
+    if (slidingWindow) {
+      return slidingWindow
+    }
+    const newSlidingWindow = new SlidingWindow({ rate: this.rate, slotProvider: this.slotProvider, key })
+    this.slidingWindows.set(key, newSlidingWindow)
+    return newSlidingWindow
+  }
+}
+
+export class SlidingWindow {
+  private readonly samplingPeriodSeconds: number
+  private readonly rateLimit: number
+  private readonly slots: Map<number, Slot> = new Map()
+  private readonly nowSeconds: () => number
+  private readonly slotProvider: SlotProvider
+  private readonly key: string
+  private rateLimitedUntil?: number
+
+  constructor({
+    rate,
+    nowSeconds,
+    slotProvider,
+    key,
+  }: {
+    rate: Rate
+    nowSeconds?: () => number
+    slotProvider: SlotProvider
+    key: string
+  }) {
+    if (rate.period < 1) {
+      throw new Error('Sampling period must be at least 1 second')
+    }
+    if (rate.requests < 0) {
+      throw new Error('Rate limit must be a positive duration')
+    }
+    this.samplingPeriodSeconds = rate.period * (rate.unit === 'hours' ? 3600 : rate.unit === 'minutes' ? 60 : 1)
+    this.rateLimit = rate.requests
+    this.nowSeconds = nowSeconds ?? (() => new Date().getTime() / 1000.0)
+    this.slotProvider = slotProvider
+    this.key = key
+  }
+
+  private getSlots(): [Slot, Slot] {
+    const now = this.nowSeconds()
+    const actualFrom = Math.floor(now / this.samplingPeriodSeconds) * this.samplingPeriodSeconds
+    const oldFrom = actualFrom - this.samplingPeriodSeconds
+
+    const oldSlot =
+      this.slots.get(oldFrom) ??
+      this.slotProvider.create({
+        fromSecond: oldFrom,
+        durationSecond: this.samplingPeriodSeconds,
+        key: this.key,
+      })
+    this.slots.set(oldFrom, oldSlot)
+    const actualSlot =
+      this.slots.get(actualFrom) ??
+      this.slotProvider.create({
+        fromSecond: actualFrom,
+        durationSecond: this.samplingPeriodSeconds,
+        key: this.key,
+      })
+    this.slots.set(actualFrom, actualSlot)
+
+    //Remove expired slots reference
+    if (this.slots.size > 2) {
+      const slots = [...this.slots.values()].sort((slot1, slot2) => slot2.fromSecond - slot1.fromSecond)
+      for (let i = 2; i < slots.length; i++) {
+        this.slots.delete(slots[i].fromSecond)
+      }
+    }
```

Comment:
Why not make this an internal function named `removeExpiredSlots`? It would document itself better and make the `getSlots` function smaller and easier to read